### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized conv/MoE model with custom float arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#364]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(80347304893125).epochs(1).log(all).stop(0.0).f(6, 9).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold. Earliest public failure: `recipe.train().run()` stalls at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (`recipe.rs` 2828-2896, NVIDIA `ptx`/`ptxas` branch) before any epoch arithmetic. The reproduction uses `recipe.model().conv(8, 1).selu().norm(batch).qi(5).1.layer(8).moe(1, [layer(8), layer(8)]).prelu().norm(batch).iq(1).m.loss(huber)` with `.f(6, 9)` arithmetic. The `.conv()` lowers through `lower_conv`, the `.layer()` through `lower_project`, and the `.moe()` through `lower_moe` (`recipe.rs` ~5419) together with quantized nodes `.qi(5).1` and `.iq(1).m`; this reaches the same native-artifact terminal check but through a `conv` + `MoE` path that no open nv0 issue represents (363 covers `res` + `MoE`; 315 covers `perc` + `MoE`; 271 covers `res` only). The loader (`data/temporal/entity_subfolders`, 24 subfolders, ~2400 rows, 1 feature) is ignored. Workload is small (width 8, one epoch, single feature) and justifies sub-second nv0 throughput; the 60-second gap is dominated by the `ptx`/`ptxas` compilation stall, not arithmetic or memory traffic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `268533dc2c93c573`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized conv/MoE model with custom float arithmetic",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.001).seed(80347304893125).epochs(1).log(all).stop(0.0).f(6, 9).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold. Earliest public failure: `recipe.train().run()` stalls at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (`recipe.rs` 2828-2896, NVIDIA `ptx`/`ptxas` branch) before any epoch arithmetic. The reproduction uses `recipe.model().conv(8, 1).selu().norm(batch).qi(5).1.layer(8).moe(1, [layer(8), layer(8)]).prelu().norm(batch).iq(1).m.loss(huber)` with `.f(6, 9)` arithmetic. The `.conv()` lowers through `lower_conv`, the `.layer()` through `lower_project`, and the `.moe()` through `lower_moe` (`recipe.rs` ~5419) together with quantized nodes `.qi(5).1` and `.iq(1).m`; this reaches the same native-artifact terminal check but through a `conv` + `MoE` path that no open nv0 issue represents (363 covers `res` + `MoE`; 315 covers `perc` + `MoE`; 271 covers `res` only). The loader (`data/temporal/entity_subfolders`, 24 subfolders, ~2400 rows, 1 feature) is ignored. Workload is small (width 8, one epoch, single feature) and justifies sub-second nv0 throughput; the 60-second gap is dominated by the `ptx`/`ptxas` compilation stall, not arithmetic or memory traffic.",
  "rationale": "Performance packet, stable `stable:true`, measured 60 s on `nv0`. Read open nv0 performance issues (271, 315, 349, 363, 322, etc.) and inspected `recipe.rs` `compile_native_artifact` (2828-2896) and lowering (`lower_conv` 5297, `lower_project` 5113, `lower_moe` 5419). The reproduction reaches the identical earliest operation (`recipe.train().run()` stall before epoch arithmetic) through a different lowering sequence (`conv` + `MoE` with `.f(6,9)`) that is not covered by any existing open issue (363 requires `.res()`, 315 requires `.perc()` and integer arithmetic, 271 has no `MoE`). The dataset loader and activation variants (`selu`/`prelu`) execute after the stall and do not alter the terminal check. Expected runtime for ~2400 rows, width 8, one epoch is well under one second; measured 60 s is materially slower and dominated by native-artifact compilation. Therefore the packet establishes a distinct earliest cause (`conv` + `MoE` quantized custom-float lowering) and must be `new`."
}
```

## Deterministic failure packet

kind=performance
id=268533dc2c93c573
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:11175 next:11176 composition:80347304893125
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-118-11175/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/entity_subfolders")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.conv(8, 1).selu().norm(batch).qi(5).1
		.layer(8)
		.moe(1, [layer(8), layer(8)]).prelu().norm(batch).iq(1).m
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(80347304893125)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Failed on Thursday, August 27, 2026 2:33 PM PDT

Commit: 8f386573e4b6979d7e4d03204602144888a2ebe6

Command:
```text
recipe test.rs
```

Reproduction script:
```rust
use recipe::*;
use std::time::{Duration, Instant};

fn main() {
	let bundle = format!("/tmp/recipe-issue-364-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/entity_subfolders")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.conv(8, 1).selu().norm(batch).qi(5).1
		.layer(8)
		.moe(1, [layer(8), layer(8)]).prelu().norm(batch).iq(1).m
		.loss(huber);
	let started = Instant::now();
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(80347304893125)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	let elapsed = started.elapsed();
	let final_loss = report.final_loss();
	println!("final_loss_finite={}", final_loss.is_finite());
	println!("elapsed_seconds={:.3}", elapsed.as_secs_f64());
	println!("within_60_seconds={}", elapsed < Duration::from_secs(60));
	assert!(final_loss.is_finite());
	assert!(elapsed < Duration::from_secs(60));
}
```

Full output:
```text
recipe: unknown command "test.rs"; expected `probe`

Recipe command line

Usage:
  recipe probe [OPTIONS]

The zero-argument probe uses the embedded theoretical seed contract, discovers
the current bare-metal machine, benchmarks every discovered device and link,
and writes an identity-keyed measured profile under the user's private cache.

Options:
  --contract PATH       Use another theoretical probe seed contract
  --profile PATH        Write/load this absolute measured-profile path
  --cuda-driver PATH    Exact CUDA Driver library candidate (repeatable)
  --hsa-runtime PATH    Exact ROCr/HSA runtime candidate (repeatable)
  --llvm-opt PATH       Exact LLVM IR verifier
  --llvm-llc PATH       Exact LLVM code generator
  --lld PATH            Exact ELF linker
  --ptxas PATH          Exact NVIDIA PTX assembler
  -h, --help             Show this help

```

Verdict: FAIL
Reason: The required public `recipe test.rs` path is absent because the `recipe` executable accepts only `probe`. The reproduction never reaches model compilation or the one-epoch nv0 run, so finite loss, completion within 60 seconds, and throughput proportionate to the stated workload remain unproven.
